### DISCUSSION TITLE
next": ^14.0.4 and next-auth:beta throwing error Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'node_modules\next\headers' imported from node_modules\next-auth\lib\index.js  Did you mean to import next/headers.js?

### DISCUSSION THREAD

**Discussion by jreyesco16** [nextauthjs/next-auth#9385] [Help]
Not sure what exactly is causing this issue but I recently upgraded from next-auth:^4.24.5 to next-auth:beta and now getting this error:

``` bash
dashboard:dev:  ⨯ Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'C:\Users\JesseCortes\Desktop\turbo-workers2\node_modules\next\headers' imported from C:\Users\JesseCortes\Desktop\turbo-workers2\node_modules\next-auth\lib\index.js
dashboard:dev: Did you mean to import next/headers.js?
dashboard:dev:     at finalizeResolution (node:internal/modules/esm/resolve:255:11)
dashboard:dev:     at moduleResolve (node:internal/modules/esm/resolve:908:10)
dashboard:dev:     at defaultResolve (node:internal/modules/esm/resolve:1121:11)
dashboard:dev:     at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:396:12)
dashboard:dev:     at ModuleLoader.resolve (node:internal/modules/esm/loader:365:25)
dashboard:dev:     at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:240:38)
dashboard:dev:     at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:85:39)
dashboard:dev:     at link (node:internal/modules/esm/module_job:84:36) {
dashboard:dev:   code: 'ERR_MODULE_NOT_FOUND',
dashboard:dev:   url: 'file:///C:/Users/JesseCortes/Desktop/turbo-workers2/node_modules/next/headers'
dashboard:dev: }
```

I would like to know if anyone else is experiencing this issue?

**Comment by dohard-ma** [comment]
I also encountered this error. When I successfully logged in using gitlab authorization and jumped to the `http://localhost:3000/api/auth/callback/gitlab?code=xxxx` address, this error was triggered.
![image](https://github.com/nextauthjs/next-auth/assets/31983261/9921f6dc-5d41-40d5-a6fb-35de74e146e6)

    "next": "14.0.4",
    "next-auth": "5.0.0-beta.4",

**Comment by tom-b001** [comment]
I'm seeing the same thing trying to access /api/auth/providers.

"next": "14.0.4",
"next-auth": "^5.0.0-beta.4",

**Comment by claudiorivera** [comment]
I've got two apps using:
```
"next": "14.0.4",
"next-auth": "5.0.0-beta.4",
```
but one uses App Router and the other uses the Pages Router. The one using App Router works like a charm.

The Pages Router version throws when I try to call `auth(context)` inside of `getServerSideProps`:
```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/claudio/Developer/my-app/node_modules/.pnpm/next-auth@5.0.0-beta.4_next@14.0.4_nodemailer@6.9.8_react@18.2.0/node_modules/next/headers' imported from /Users/claudio/Developer/my-app/node_modules/.pnpm/next-auth@5.0.0-beta.4_next@14.0.4_nodemailer@6.9.8_react@18.2.0/node_modules/next-auth/lib/actions.js
Did you mean to import next@14.0.4_react-dom@18.2.0_react@18.2.0/node_modules/next/headers.js?
    at new NodeError (node:internal/errors:405:5)
    at finalizeResolution (node:internal/modules/esm/resolve:324:11)
    at moduleResolve (node:internal/modules/esm/resolve:943:10)
    at defaultResolve (node:internal/modules/esm/resolve:1129:11)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:835:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
    at link (node:internal/modules/esm/module_job:76:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

**Comment by ivomasterche** [comment]
"next": "14.0.4",
"next-auth": "5.0.0-beta.4",

In my case it's an App router.  It throws the error only when in the page (or in its child components) is an import connected to next-auth. 
My home page works ok, even my login page works ok, but after the login when a component for the authorization is involved it throws the error.

**Comment by Joel-PeakMetrics** [comment]
For me it is the signin page. When calling getProviders(), it throws this error.

**Comment by sandercoffee** [comment]
any updates? same error using vitest test library

**Comment by zhaozhonglei** [comment]
I also got this error, Cannot find module 'node_modules/next/server' imported from node_modules/next-auth/lib/env.js .
"next": "14.1.0",
"next-auth": "^5.0.0-beta.13",

**Reply by Natanael-Exe** [reply]
same issue

**Reply by voices-in-tech** [reply]
Same issue - working fine on execution but not through vitest

**Reply by md-coops** [reply]
Same issue with my environment set up. next-auth@beta, Next14 with partial migration to next router.

**Comment by oberlinstands** [comment]
Bump!

**Comment by renatoastra** [comment]
I got this error too, im using pages dir,

` "next": "^14.1.0"`
 `"next-auth": "^5.0.0-beta.13"`

**Comment by fisher-** [comment]
Same here:
`"next": "14.1.2",`
`"next-auth": "5.0.0-beta.15",`

**Comment by tomislavhoman** [comment]
Same:
```
"next": "14.1.2",
"next-auth": "^5.0.0-beta.4",
```

**Comment by OpenTextASOWSM** [comment]
did import of auth/core from NPM, didn't help on its own. 

Ok looks like some naming issues with not finding the .js versions of these files in node_modules

node_modules/next-auth/lib/env.js

import {NextRequest} from "next/server.js";
import { setEnvDefaults as coreSetEnvDefaults } from "@auth/core";

-from-
import {NextRequest} from "next/server";
import { setEnvDefaults as coreSetEnvDefaults } from "@auth/core";

node_modules/next-auth/lib/index.js

-to-
import { Auth } from "@auth/core";
import { headers } from "next/headers.js";
import { NextResponse } from "next/server.js";
import { reqWithEnvURL } from "./env.js";
import { createActionURL } from "./actions.js";

node_modules/next-auth/lib/actions.js

-to-
import { Auth, raw, skipCSRFCheck } from "@auth/core";
import { headers as nextHeaders, cookies } from "next/headers.js";
import { redirect } from "next/navigation.js";

after making changes restarting npm run dev

I then hit errors with the /api/auth/[...nextauth] but that's different error.

**Reply by OpenTextASOWSM** [reply]
Looking the auth.js core  seems like they have other places they've had to refer to imports with .js as well. Perhaps different projects work.

I have tsconfig.json with   "allowJs": true,  so perhaps others have this issue also due to this or it should be fixed in the core library.

**Reply by OpenTextASOWSM** [reply]
reported bug - hope that's the right process - [Issue 10282](https://github.com/nextauthjs/next-auth/issues/10282)

**Reply by OpenTextASOWSM** [reply]
Ok that was rejected and I've now filed a pull request and a new [issue ](https://github.com/nextauthjs/next-auth/issues/10284)
https://github.com/nextauthjs/next-auth/pull/10283

**Reply by OpenTextASOWSM** [reply]
If you still have this an can share reproductions please reply to the issue.

**Comment by employee451** [comment]
For me, this was fixed by transpiling `next-auth` in `next.config.js`

```js
{
  transpilePackages: ['next-auth'],
}
```

**Reply by sim391calado** [reply]
It also fixed it for me, but wondering if there is a better solution.

**Reply by MrGiddz** [reply]
This fixed it for me too. But then I started getting another error that said "⨯ Error: `headers` was called outside a request scope. Read more: https://nextjs.org/docs/messages/next-dynamic-api-wrong-context"

I am new to next.js and in my case I was calling the "auth" method from "pages/lib/current-profile-pages.ts"

I was going through "node_modules/next-auth/src/index.tsx" then I figured I had to pass 
`(req: NextApiRequest, res: NextApiResponse)` as parameters inside the auth method.

so I did this 

```
export const currentProfilePages = async (
  req: NextApiRequest,
  res: NextApiResponse
) => {
  const session = await auth(req, res);
  const user = session?.user;
```

**Reply by gnomefin** [reply]
it worked using that transpile config, none of maintainer here can explain which part causing this issue, tho

I mean I know that this is not completely a major issue, however, the backward compatible between the next auth and nextJS seems not quite good for this beta version.

Thanks! @employee451

**Reply by AndreyMarchuk** [reply]
fixed it for me too
I haven't had any issues with `beta.20` but started getting an error with `beta.22` and newer versions

**Reply by frazie** [reply]
> For me, this was fixed by transpiling `next-auth` in `next.config.js`
> 
> ```js
> {
>   transpilePackages: ['next-auth'],
> }
> ```

after spending hours on hours trying to trace all possible points of errors this worked!!! Thank you!!

**Reply by IamWills** [reply]
it doesn't work for me.

**Reply by frazie** [reply]
I think the fix also depends on how you set up your project

**Comment by timmeade** [comment]
This did not fix it for me.  I have been fighting this for a few days and am really starting to not like this package anymore.   Seems like a pretty big deal if i cannot run the example code.

**Reply by liufansizhe** [reply]
+1

**Comment by OpenTextASOWSM** [comment]
Once I personally resolved this I hit other errors and dropped using the beta release and used the v4 mainstream release. I think if you are building anything right now it is probably safer to use that version and worrying about upgrading if needed. Other than rebranding not sure it has that big of a difference once I got my code working.

**Comment by nazargladish** [comment]
Encountered the same issue

/api/socket/messages.ts
<img width="656" alt="Screenshot 2024-03-30 at 19 19 19" src="https://github.com/nextauthjs/next-auth/assets/157213865/2f2c0780-37a6-4e9a-9c04-03304c67c06f">

<img width="689" alt="Screenshot 2024-03-30 at 19 20 54" src="https://github.com/nextauthjs/next-auth/assets/157213865/0e45275b-7726-4439-8680-8be88c43de8e">

**Reply by OpenTextASOWSM** [reply]
what are your imports in this file? What about in the env.js?

**Reply by zenovak** [reply]
I have the same exact issue, following the documentations. Trying to access the builtin login page throws this error:
http://localhost:3000/api/auth/signin

pages/api/auth/[...nextauth].js
```js
import NextAuth from "next-auth";
import GithubProvider from "next-auth/providers/github";


export const authOptions = {
  // Configure one or more authentication providers
  providers: [
    GithubProvider({
      clientId: process.env.GITHUB_ID,
      clientSecret: process.env.GITHUB_SECRET,
    }),
  ],

  // Here is where you add other configuration objects like sessions etc:
}

export default NextAuth(authOptions)
```

**Comment by lb-joschabombien** [comment]
Any news on this issue? I've done the changes of https://github.com/nextauthjs/next-auth/pull/10304 locally and got it working for me.

**Reply by OpenTextASOWSM** [reply]
They are saying it's an issue in next.js and they are waiting that fix apparently - https://github.com/vercel/next.js/pull/64529

**Comment by LZL0** [comment]
Same issue.

**Comment by alexrandaccio** [comment]
I have the same issue

Error [ERR_MODULE_NOT_FOUND]: Cannot find module '//node_modules/next/server' imported from //node_modules/next-auth/lib/env.js
Did you mean to import "next/server.js"

"next": "^14.2.3",
"next-auth": "^5.0.0-beta.17",

**Comment by davidgomes** [comment]
FYI I have had the same issue, migrating back to NextAuth.js 4 is the only thing that fixed the issue for me. This is completely fair and expected as Auth.js 5 is still in beta.

**Reply by ajayraho** [reply]
Thanks!
Guess this is the only working solution.

**Comment by OpenTextASOWSM** [comment]
The permanent fix seems to be making progress in next.js - https://github.com/vercel/next.js/pull/64529 partially approved. Some other revisions requested 3 days ago for some files to work with JS and TS. Hopefully that's done relatively soon in comparison to how long this has been experienced.

**Comment by bavjackson** [comment]
@sandercoffee @voices-in-tech 
For anyone who finds this while trying to get vitest running, try adding the following lines to your `vitest.config.js/ts`.

```
export default defineConfig({
  // other config goes here
  test: {
    // other test config
    server: {
      deps: {
        inline: ['next']
      }
    }
  }
}
```

**Reply by voices-in-tech** [reply]
Hi @bavjackson many thanks for sharing

Yeah I'd seen this one mentioned somewhere as well 

Not sure if its my specific config but I get this error when trying the fix suggested

```
Error: `headers` was called outside a request scope. Read more: https://nextjs.org/docs/messages/next-dynamic-api-wrong-context
 ❯ getExpectedRequestStore node_modules/next/src/client/components/request-async-storage.external.ts:31:9
 ❯ Module.headers node_modules/next/src/client/components/headers.ts:36:33
 ❯ Module.<anonymous> node_modules/next-auth/lib/index.js:95:53
     93|         if (args[0] instanceof Request) {
     94|             // middleware.ts inline
     95|             // export { auth as default } from "auth"
       |                                                     ^
     96|             const req = args[0];
     97|             const ev = args[1];
```

**Comment by jcdogo** [comment]
This commenter mentions that you can transpile nextauth: https://github.com/nextauthjs/next-auth/pull/10304#issuecomment-2077546742

In `next.config.js`
`
{
transpilePackages: ['next-auth'],
}
`
It allowed me to build the project again successfully. None of the other options in this thread worked for me except this one.

**Reply by fbarbare** [reply]
This is what worked for me and is much simpler

**Reply by thathurtabit** [reply]
Thanks @jcdogo !

**Comment by KyryloSWAN** [comment]
I renamed `src/pages/api/auth/[...nextauth].ts` to `app/api/auth/[...nextauth]/route.ts` and the error went away.

"next": "15.1.6"
"next-auth": "^5.0.0-beta.25"

**Reply by ricleedo** [reply]
I met same error with:
"next": "15.1.6"
"next-auth": "^5.0.0-beta.25"

I tried your method, but it throws this error: 
Error: Catch-all must be the last part of the URL.
    at Array.forEach (<anonymous>)

**Comment by TephrosisDEV** [comment]
I have the same error with:
"next": "15.2.3",
"next-auth": "5.0.0-beta.25"

- I tried all next version starting from 15.1.3 and up
- I tried all beta's from beta.19 and up

However I'm using `pnpm@10.11.0`, so maybe it has something to do with that as well?

```
⨯ [Error: Cannot find module '/Users/<user>/Documents/Code/<project>/node_modules/.pnpm/next-auth@5.0.0-beta.25_next@15.2.3_@opentelemetry+api@1.9.0_babel-plugin-macros@3.1.0__ba3d12a5f30fa64ccfcbf0a757a6022e/node_modules/next/server' imported from /Users/<user>/Documents/Code/<project>/node_modules/.pnpm/next-auth@5.0.0-beta.25_next@15.2.3_@opentelemetry+api@1.9.0_babel-plugin-macros@3.1.0__ba3d12a5f30fa64ccfcbf0a757a6022e/node_modules/next-auth/lib/env.js
Did you mean to import "next/server.js"?] {
  type: 'Error',
  code: 'ERR_MODULE_NOT_FOUND',
  url: 'file:///Users/<user>/Documents/Code/<project>/node_modules/.pnpm/next-auth@5.0.0-beta.25_next@15.2.3_@opentelemetry+api@1.9.0_babel-plugin-macros@3.1.0__ba3d12a5f30fa64ccfcbf0a757a6022e/node_modules/next/server',
  page: '/nl-BE/admin/partners'
}
```
